### ISSUE TITLE
tab out -> segfault

### ISSUE THREAD

**Issue by Turmfalke2** [ValveSoftware/Dota-2#378]
hi,
I just tried to do the last hit practise mode (dota_tutorial_start lasthit), but noticed that it failed to properly load the map and I never got the hero selection screen. So I switched to another tag (using herbstluftwm) googled and learned that apparently the last hit challenge isn't implemented in reborn. A moment later I heard a throne explode and the game crashed.
In dmesg it states:
[3125867.392390] GLRenderThread[6965]: segfault at 7f81f3bb000c ip 00007f81fe5f9cb5 sp 00007f81f3bad400 error 4 in libnvidia-glcore.so.352.21[7f81fcd89000+19fb000]

I am currently using x11-drivers/nvidia-drivers-352.21, could look into updating the driver later today.

**Comment by davidw-valve** [comment]
To be clear this only happens if you use a console command to access it? There is no navigable path to it through the gui?

**Comment by Turmfalke2** [comment]
I don't think it was related to the last hit practise mode, that is just how I encountered it. And yes there is no navigable path to the none existent practise mode. I have tried to reproduce it, but wasn't able.

**Comment by danslo** [comment]
I get this too:

`[90172.790620] GLRenderThread[20349]: segfault at 5606d35d9160 ip 00007fc2d74d7958 sp 00007fc2cc28d620 error 4 in libnvidia-glcore.so.355.06[7fc2d649a000+126e000]`.

Driver is 355.06.

**Comment by Turmfalke2** [comment]
Got this again yesterday. I was tapped out while we were searching for a ranked party game, the moment I heard the game is ready sound the game crashed.
[106723.124815] GLRenderThread[22396]: segfault at 7f5b611e9d80 ip 00007f5f7593bf8c sp 00007f5f6aca5640 error 4 in libnvidia-glcore.so.352.21[7f5f74172000+19fb000]

**Comment by danginsburg** [comment]
Does running with -gl_disable_buffer_storage help?  Can you post full system spec from a crash?  It would also be helpful if you could post a minidump.

**Comment by danslo** [comment]
I'll try that option and see if it prevents crashes. It's not exactly reproducible either, it just randomly happens between tabbing. 

Running proprietary linux driver 355.11 (latest as of writing this), however it has also happened on other versions of that driver such as 355.06.

How do I obtain a minidump?

**Comment by danginsburg** [comment]
If you go to Steam -> Help -> System Information you will see a list of recent minidumps.

**Comment by danslo** [comment]
Managed to crash it again, even with `-gl_disable_buffer_storage`.

dmesg: `[78987.698081] GLRenderThread[13803]: segfault at fffffff8edca5ca0 ip 00007fbfec1fb728 sp 00007fbfe07ec620 error 5 in libnvidia-glcore.so.355.11[7fbfeb1bd000+126f000]`

```
Processor Information:
    Vendor:  GenuineIntel
    CPU Family:  0x6
    CPU Model:  0x1a
    CPU Stepping:  0x5
    CPU Type:  0x0
    Speed:  3068 Mhz
    8 logical processors
    4 physical processors
    HyperThreading:  Supported
    FCMOV:  Supported
    SSE2:  Supported
    SSE3:  Supported
    SSSE3:  Supported
    SSE4a:  Unsupported
    SSE41:  Supported
    SSE42:  Supported

Network Information:
    Network Speed:  

Operating System Version:
    Ubuntu Wily Werewolf (development branch) (64 bit)
    Kernel Name:  Linux
    Kernel Version:  4.2.0-14-generic
    X Server Vendor:  The X.Org Foundation
    X Server Release:  11702000
    X Window Manager:  GNOME Shell
    Steam Runtime Version:  steam-runtime-release_2015-06-12

Video Card:
    Driver:  NVIDIA Corporation GeForce GTX 460/PCIe/SSE2

    Driver Version:  4.5.0 NVIDIA 355.11
    OpenGL Version: 4.5
    Desktop Color Depth: 24 bits per pixel
    Monitor Refresh Rate: 60 Hz
    VendorID:  0x10de
    DeviceID:  0xe22
    Number of Monitors:  2
    Number of Logical Video Cards:  1
    Primary Display Resolution:  1920 x 1080
    Desktop Resolution: 3840 x 1080
    Primary Display Size: 20,08" x 11,42"  (23,07" diag)
                                            51,0cm x 29,0cm  (58,6cm diag)
    Primary Bus: PCI Express 16x
    Primary VRAM: 1024 MB
    Supported MSAA Modes:  2x 4x 8x 16x 

Sound card:
    Audio device: USB Mixer

Memory:
    RAM:  5958 Mb

Miscellaneous:
    UI Language:  English
    LANG:  en_US.UTF-8
    Microphone:  Not set
    Total Hard Disk Space Available:  106507 Mb
    Largest Free Hard Disk Block:  12449 Mb

Installed software:

Recent Failure Reports:
    Mon Oct 12 17:48:20 2015 GMT: file ''/tmp/dumps/crash_20151012190846_1.dmp'', upload yes: ''Discarded=1''
```

The crash dump has been uploaded [here](https://github.com/danslo/Dota-2/raw/master/crash_20151012190846_1.dmp).

**Comment by danginsburg** [comment]
Thank you for the minidump.  So unfortunately this is crashing in a thread spawned by the NVIDIA driver when the game issues a swap buffers.  You could try editing ~/.steam/steam/steamapps/common/dota 2 beta/game/dota.sh and modifying this line:

export __GL_THREADED_OPTIMIZATIONS=1

to:

export __GL_THREADED_OPTIMIZATIONS=0

If this does work, keep in mind that dota.sh will get overwritten on the next update.

**Comment by danslo** [comment]
That does seem to have solved it - no crashes so far, though does (expectedly) result in noticably degraded performance especially when a lot is happening on the screen (earthshaker or death prophet ultimates for example).

**Comment by danginsburg** [comment]
OK, this very strongly points to a bug with driver multithreading in the NVIDIA driver.  We will follow up with NVIDIA.

**Comment by emnoor** [comment]
It happens with me too. Just adding this comment to say that it is may be not a NVIDIA related issue as it happens for me too.

Distro: Arch Linux
DE: Gnome
CPU: Intel i5-4770
GPU: (cpu-builtin) Intel HD 4600

It happens randomly with overlays too (like gnome's volume overlay). I have noticed that, if it does not crash for a tabout/overlay, then it will not crash for any following tabouts until you close the game. Crash happens again for next start of the game (randomly).

**Comment by danslo** [comment]
> We will follow up with NVIDIA.

Thank you. If there is anything else I can assist with, please let me know.

**Comment by danslo** [comment]
This problem got a lot worse after installing version `358.09` of the proprietary nvidia drivers. Just crashes on _any_ alt tab or overlay (like gnome volume) now.

**Comment by danslo** [comment]
Same goes for `358.16`.

**Comment by catichenor** [comment]
Hi @danslo, I'm taking a look at this bug and I'm having some problems reproducing it. Could you upload an nvidia-bug-report log (`sudo nvidia-bug-report.sh`)? 

Also, I'm attempting to reproduce in standard Ubuntu 14.04, are you using a different setup?

**Comment by danslo** [comment]
@catichenor It's 15.10 but I've been having the issue in older (LTS) releases. I'm using Gnome (3.18.2, though older versions of 3.x had same problem) instead of Unity though.

**Comment by danslo** [comment]
Nvidia bug report can be found [here](https://raw.githubusercontent.com/danslo/Dota-2/master/nvidia-bug-report.log.gz).

**Comment by danslo** [comment]
This has been solved for me with the `361.16` driver. Thanks a lot @catichenor .

**Comment by danginsburg** [comment]
Great, glad to hear it!  Closing.
